lib/Sema/ConstraintSystem.cpp

```diff
@@ -1761,8 +1761,17 @@ static void validateInitializerRef(ConstraintSystem &cs, ConstructorDecl *init,
             cs, baseType, init, locator));
         return;
       }
+    } else if (baseType->is<AnyMetatypeType>() &&
+               baseType->getMetatypeInstanceType()
+                   ->getWithoutParens()
+                   ->isAnyExistentialType()) {
+      (void)cs.recordFix(AllowInvalidInitRef::onProtocolMetatype(
```

Comment:
You can integrate all of this into previous if statement. 

Also `P.Type` and `P.Protocol` types used as base for initializer should produce different errors - first should say that `.init` should be used explicitly like it does right now, the latter one should stay that protocol couldn't be instantiated.

Reply:
@xedin Done.

I am not sure about that, as for example:

```
let _ = p().bar(1)
```

in this case, the instance type is an existential type, so it doesn't meet the checks for `onNonConstMetatype` which is why this new diagnostic is emitted.